neural_compressor/adaptor/tf_utils/smooth_quant_scaler.py

```diff
@@ -0,0 +1,148 @@
+        """Apply scaling to weights and activations based on the maximum values per channel.
+
+        Args:
+            max_vals_per_channel (dict): A dictionary containing the maximum values per channel for each input node.
+            sq_weight_tensors (dict): A dictionary containing the weight tensors for each input node.
+            sq_weights_nodes (dict): A dictionary containing the constant nodes for each input node.
+            sq_weight_node_names (dict): A dictionary from weight node name to the its concrete output node name.
+        
+        Returns:
+            tuple: A tuple containing the modified model and a list of the inserted multiplication nodes.
+        """
+        if self.scales_per_op:
+            # 1. obtain the smooth scale per op
+            # 2. adjust weight
+            # 3. adjust activation
+            for idx, input_node_name in enumerate(max_vals_per_channel):
+                A_max_per_in_channel = max_vals_per_channel[input_node_name]
+                W_lst = sq_weight_tensors[input_node_name]  # VQK weight value
+                # Use the const nodes before to get weight values, VQK ReadVariable
+                W_const_node_lst = sq_weights_nodes[input_node_name]
+                # W_node_lst = sq_node_names[input_node_name]
+                # Get the concrete weight node as the output of Mul insertion, QKV ReadVariable
+                for w_i, W in enumerate(W_lst):
+                    if len(W.shape) == 4:
+                        # https://www.tensorflow.org/api_docs/python/tf/nn/conv2d
+                        # weight: [filter_height, filter_width, in_channels, out_channels]
+                        # activation: NHWC, also batch_shape + [in_height, in_width, in_channels]
+                        tensor = np.abs(np.transpose(W, [0, 1, 3, 2]))
+                        # reduce weight max to (in_channel, ), aligned with activation max
+                        W_max_per_in_channel = np.max(np.reshape(tensor, (-1, tensor.shape[-1])), axis=0)
+                    elif len(W.shape) == 2: # matmul
+                        # reduce weight max to (in_channel, ), aligned with activation max
+                        tensor = np.abs(W)
+                        W_max_per_in_channel = np.max(tensor, axis=1)
+                    else:
+                        assert False, "not supported"
+                    cur_const_node = W_const_node_lst[w_i]
+                    try:
+                        scale = np.power(A_max_per_in_channel, self.alpha) /  \
+                                np.power(W_max_per_in_channel, (1-self.alpha))
+                    except ValueError as e:
+                        logger.info(e)
+                        logger.info("Skip smoothing the node: {}".format(cur_const_node.name))
+                        continue
+                    # clip the scales that are too small
+                    scale = np.clip(scale, a_min=1e-2, a_max=1e8)
```

Comment:
may I know whether there are some models that benefited from this setting? We may apply this setting to other backends.

Reply:
Hi, I'm not sure which line of code you are refering to that makes some "benefit"?

If you mean the try-catch clock from line 130-136, we catch the exception because I remember that in the resnet there is a const node that will cause "operands could not be broadcast together with shapes...", which means that the two inputs' shape do not match for the computation. Also at that time we check that intel tensorflow does not support optimization on that specific setting so we just skip that.

If you mean the line 138 clipping of the scales, I have no idea on that line of code :) because I read about PyTorch and ONNX backend sq code and find this line. I think it may skip the "divide by zero" problem (but for numpy I think it only will raise a Warning rather than an Error), so I do not really think this line make sense.

If you mean the line 139-142, this few lines (need uncommented) is useful for the distilbert. I use that because I think smoothquant is not really a silver bullet for restoring accuracy. I have done some experiments to see that there are some cases when the MeanSquareError increases after using smoothquant, compared to not using sq (we expect MSE to be lower after using sq). The paper assumes that 1) the activation >> weight for some channels 2) for the same channel, variance is relatively small. However, I think this is too ideal for arbitrary models. For example, I find that for one distilbert model I have to uncomment 139-142 to get the accuracy correct. That is just my experience and it may not be correct, but it may works if you meet relevant problems for other backends.

Reply:
Thanks for pointing out. I remember that min value does not make some clear effect or benefit. Maybe I just forgot to edit it back. I will check and fix that.